kevm-pyk/src/kevm_pyk/kdist/_kdist.py

```diff
@@ -50,118 +52,158 @@ def _dist_dir() -> Path:
+        res: dict[str, dict[str, Target]] = {}
+        for plugin in plugins:
+            plugin_name = plugin.name
 
-    @property
-    def dir(self) -> Path:
-        return _DIST_DIR / self.plugin / self.target
+            if not _valid_id(plugin_name):
+                _LOGGER.warning(f'Invalid plugin name, skipping: {plugin_name}')
+                continue
 
+            _LOGGER.info(f'Loading kdist plugin: {plugin_name}')
+            module_name = plugin.value
+            try:
+                _LOGGER.info(f'Importing module: {module_name}')
+                module = importlib.import_module(module_name)
+            except Exception:
+                _LOGGER.error(f'Module {module_name} cannot be imported', exc_info=True)
+                continue
 
-def plugins() -> dict[str, list[str]]:
-    return {plugin: list(targets) for plugin, targets in _PLUGINS.items()}
+            res[plugin_name] = TargetCache._load_targets(module)
 
+        return res
 
-def targets() -> list[str]:
-    return [TargetId(plugin, target).fqn for plugin, targets in plugins().items() for target in targets]
+    @staticmethod
+    def _load_targets(module: ModuleType) -> dict[str, Target]:
+        if not hasattr(module, '__TARGETS__'):
+            _LOGGER.warning(f'Module does not define __TARGETS__: {module.__name__}')
+            return {}
 
+        targets = module.__TARGETS__
 
-def resolve(target_fqn: str) -> TargetId:
-    segments = target_fqn.split('.')
-    if len(segments) != 2:
-        raise ValueError(f'Expected fully qualified target name, got: {target_fqn!r}')
+        if not isinstance(targets, Mapping):
+            _LOGGER.warning(f'Invalid __TARGETS__ attribute: {module.__name__}')
+            return {}
 
-    plugin, target = segments
+        res: dict[str, Target] = {}
+        for key, value in targets.items():
+            if not isinstance(key, str):
+                _LOGGER.warning(f'Invalid target key in {module.__name__}: {key!r}')
+                continue
 
-    if not _valid_id(plugin):
-        raise ValueError(f'Invalid plugin identifier: {plugin!r}')
+            if not _valid_id(key):
+                _LOGGER.warning(f'Invalid target name (in {module.__name__}): {key}')
+                continue
 
-    if not _valid_id(target):
-        raise ValueError(f'Invalid target identifier: {target!r}')
+            if not isinstance(value, Target):
+                _LOGGER.warning(f'Invalid target value in {module.__name__} for key {key}: {value!r}')
+                continue
 
-    _plugins = plugins()
+            res[key] = value
 
-    try:
-        targets = _plugins[plugin]
-    except KeyError:
-        raise ValueError(f'Undefined plugin: {plugin}') from None
+        return res
 
-    if not target in targets:
-        raise ValueError(f'Plugin {plugin} does not define target: {target}')
 
-    return TargetId(plugin, target)
+_DIST_DIR: Final = _dist_dir()
+_CACHE: Final = TargetCache.load()
```

Comment:
Does this mean we're loading the target cache every time this process starts up? Is that only for calls to `kevm-dist` or also for calls to `kevm ...`?

Reply:
The cache is populated on the first import to `kdist`. Although there's some extra processing involved, I don't think it's significantly more overhead then importing `plugin` directly.

If it turns out to be a bottleneck, we can separate "producers" like `kdist.build` from "consumers" like `kdist.get` so that code that's only interested in accessing definitions don't have to wait for the cache to be populated, though It's still useful for error handling.